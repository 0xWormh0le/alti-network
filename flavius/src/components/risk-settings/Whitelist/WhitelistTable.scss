@import 'styles/_shared.scss';

.WhitelistTable {
  &__delete {
    text-transform: uppercase;
    margin-left: auto;
  }
}

.WhitelistItem {
  margin-bottom: $SMALL_SPACING;

  &.Table {
    width: calc(100% - 1px); // this is to fix the right border disappear bug
  }
}
