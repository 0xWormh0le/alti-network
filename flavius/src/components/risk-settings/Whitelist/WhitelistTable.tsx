import React from 'react'
import cx from 'classnames'
import ContentLoader from 'react-content-loader'
import { ModernTable } from '@altitudenetworks/component-library'
import Button from 'components/elements/Button'
import EntityCountPagination from 'components/elements/EntityCountPagination'
import UI_STRINGS from 'util/ui-strings'
import { retrieveFromStorage } from 'util/storage'
import './WhitelistTable.scss'

export interface WhitelistTableProps<TResponse, TData> {
  className?: string
  column: string
  pageNumber: number
  isLoading: boolean
  response: TResponse | null
  onDelete: (val: TData) => void
  mapResponseToData: (value: TResponse) => TData[]
  render: (value: TData) => React.ReactNode
  onPageChange: (value: number) => void
  useAllowDelete?: boolean
}

const storedDomains = retrieveFromStorage<string[]>('domains')

const WhitelistTable = <
  TResponse extends WhitelistedDomainsResponse | WhitelistedAppsResponse | InternalDomainsResponse,
  TData extends RiskSettingsDomain | RiskSettingsApp
>({
  className,
  column,
  mapResponseToData,
  pageNumber,
  onPageChange,
  isLoading,
  response,
  onDelete,
  useAllowDelete,
  render
}: WhitelistTableProps<TResponse, TData>) => (
  <ModernTable
    isLoading={isLoading}
    loadingComponent={<WhitelistTableLoading />}
    className={cx('WhitelistItem', className)}
    items={response ? mapResponseToData(response) : []}
    fields={[column]}
    scopedSlots={{ [column]: render }}
    actions={(val) => {
      const domainVal = (val as RiskSettingsDomain)?.domain
      if (useAllowDelete) {
        if (!(val as RiskSettingsDomain)?.allowDelete) {
          return null
        }
      } else {
        if (domainVal && storedDomains?.includes(domainVal)) return null
      }

      return (
        <Button
          className='WhitelistTable__delete'
          action='submit'
          onClick={() => onDelete(val)}
          text={UI_STRINGS.BUTTON_LABELS.DELETE}
        />
      )
    }}
    noResultsMessage={UI_STRINGS.ERROR_MESSAGES.NO_ENTRIES}>
    <EntityCountPagination
      totalEntityCount={response ? response.pageSize * response.pageCount : 0}
      entityCount={response ? response.pageSize : 0}
      onPageChange={onPageChange}
      pageCount={response?.pageCount || 0}
      pageNumber={Number(pageNumber)}
      pageSize={response?.pageSize || 0}
    />
  </ModernTable>
)

const WhitelistTableLoading = () => (
  <ContentLoader
    backgroundColor='#F0F0F0'
    foregroundColor='#F7F7F7'
    height={400}
    width='100%'
    uniqueKey='AppSpotlightAppImage__title'>
    <rect x={0} y={0} width='100%' height={300} rx={4} ry={4} />
  </ContentLoader>
)

export default WhitelistTable
