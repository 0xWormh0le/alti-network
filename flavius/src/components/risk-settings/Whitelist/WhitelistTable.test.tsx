import React from 'react'
import { render } from '@testing-library/react'
import { WhitelistTableProps, default as WhitelistTable } from './WhitelistTable'
import { noop } from 'lodash'

describe('WhitelistTable', () => {
  it('render correctly', () => {
    const props: WhitelistTableProps<WhitelistedDomainsResponse, RiskSettingsDomain> = {
      mapResponseToData: (data: WhitelistedDomainsResponse) => data.settings.whitelistedDomains,
      pageNumber: 1,
      onPageChange: noop,
      isLoading: false,
      column: 'Domain',
      response: {
        settings: {
          whitelistedDomains: [
            {
              domain: 'alice.com'
            },
            {
              domain: 'bob.com'
            },
            {
              domain: 'carl.com'
            }
          ]
        },
        pageNumber: 1,
        pageSize: 10,
        pageCount: 1,
        pageCountCacheTTL: 1,
        pageCountLastUpdated: 1
      } as WhitelistedDomainsResponse,
      onDelete: noop,
      render: (value: RiskSettingsDomain) => (
        <span className='Whitelist__table-item'>{(value as RiskSettingsDomain).domain}</span>
      )
    }
    const { container } = render(<WhitelistTable {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('renders Delete button only when allowDelete field is specified for internal domains', () => {
    const props: WhitelistTableProps<InternalDomainsResponse, RiskSettingsDomain> = {
      mapResponseToData: (data: InternalDomainsResponse) => data.settings.internalDomains,
      pageNumber: 1,
      onPageChange: noop,
      isLoading: false,
      column: 'Domain',
      response: {
        settings: {
          internalDomains: [
            {
              domain: 'alice.com',
              allowDelete: true
            },
            {
              domain: 'bob.com',
              allowDelete: false
            },
            {
              domain: 'carl.com',
              allowDelete: false
            }
          ]
        },
        pageNumber: 1,
        pageSize: 10,
        pageCount: 1,
        pageCountCacheTTL: 1,
        pageCountLastUpdated: 1
      } as InternalDomainsResponse,
      onDelete: noop,
      useAllowDelete: true,
      render: (value: RiskSettingsDomain) => (
        <span className='Whitelist__table-item'>{(value as RiskSettingsDomain).domain}</span>
      )
    }
    const { container } = render(<WhitelistTable {...props} />)
    expect(container).toMatchSnapshot()
  })
})
