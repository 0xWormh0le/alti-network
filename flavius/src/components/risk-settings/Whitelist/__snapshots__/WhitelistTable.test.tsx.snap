// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhitelistTable render correctly 1`] = `
<div>
  <div
    class="Table WhitelistItem"
  >
    <table
      class="Table__table"
    >
      <thead>
        <tr
          class="Table__heading"
        >
          <th
            class="Table__heading-cell"
          >
            <span>
              Domain
            </span>
          </th>
          <th
            class="Table_heading-cell"
          />
        </tr>
      </thead>
      <tbody>
        <tr
          class="Table__row"
        >
          <td
            class="Table__cell"
          >
            <span
              class="Whitelist__table-item"
            >
              alice.com
            </span>
          </td>
          <td
            class="Table__cell"
          >
            <button
              class="Button WhitelistTable__delete Button--pagination"
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr
          class="Table__row"
        >
          <td
            class="Table__cell"
          >
            <span
              class="Whitelist__table-item"
            >
              bob.com
            </span>
          </td>
          <td
            class="Table__cell"
          >
            <button
              class="Button WhitelistTable__delete Button--pagination"
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr
          class="Table__row"
        >
          <td
            class="Table__cell"
          >
            <span
              class="Whitelist__table-item"
            >
              carl.com
            </span>
          </td>
          <td
            class="Table__cell"
          >
            <button
              class="Button WhitelistTable__delete Button--pagination"
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="Table__children"
    >
      <div
        class="EntityCountPagination"
      >
        <div />
        <div
          class="EntityCountPagination__pagination-info"
        >
          <div
            class="EntityCountPagination__page-size EntityCountPagination__items-count"
          >
            <span>
              Displaying 10 Items Per Page
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`WhitelistTable renders Delete button only when allowDelete field is specified for internal domains 1`] = `
<div>
  <div
    class="Table WhitelistItem"
  >
    <table
      class="Table__table"
    >
      <thead>
        <tr
          class="Table__heading"
        >
          <th
            class="Table__heading-cell"
          >
            <span>
              Domain
            </span>
          </th>
          <th
            class="Table_heading-cell"
          />
        </tr>
      </thead>
      <tbody>
        <tr
          class="Table__row"
        >
          <td
            class="Table__cell"
          >
            <span
              class="Whitelist__table-item"
            >
              alice.com
            </span>
          </td>
          <td
            class="Table__cell"
          >
            <button
              class="Button WhitelistTable__delete Button--pagination"
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr
          class="Table__row"
        >
          <td
            class="Table__cell"
          >
            <span
              class="Whitelist__table-item"
            >
              bob.com
            </span>
          </td>
          <td
            class="Table__cell"
          />
        </tr>
        <tr
          class="Table__row"
        >
          <td
            class="Table__cell"
          >
            <span
              class="Whitelist__table-item"
            >
              carl.com
            </span>
          </td>
          <td
            class="Table__cell"
          />
        </tr>
      </tbody>
    </table>
    <div
      class="Table__children"
    >
      <div
        class="EntityCountPagination"
      >
        <div />
        <div
          class="EntityCountPagination__pagination-info"
        >
          <div
            class="EntityCountPagination__page-size EntityCountPagination__items-count"
          >
            <span>
              Displaying 10 Items Per Page
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
